Expands stabbing eyes to all sharp items

emmanuelbassil opened the pull request:
🆑 emmanuelbassil
tweak: Can now stab eyes with all tiny, small, and normal sized sharp items.
tweak: It is now possible to damage glass/mask items if eye-stabbing with sufficient force. This is affected by CQC skill, making it easier to break glasses with higher skill. It is not possible to break hats that cover eyes (e.g riot helmets) in this way however, no matter the force.
tweak: It is now possible to miss while trying to stab someone's eyes.
/🆑 

Another nice fix I found while doing all my attacks()
Made eyestab properly boolean since it's failure/success now affects whether the code goes onto the next use_*
Removed two attacks() since they're now processed at use_weapon level, made the remaining one boolean.
Added force to pens to match all the other sharp items; slightly weaker than screwdrivers. (Force = 2 instead of 3).

emmanuelbassil commented:
Right so I almost lost all my work; lesson learned.
Merge conflicts solved; can also now break glasses and stab eyes depending on force and CQC skill. 
Cannot break any helmets/masks that cover eyes like riot helmets no matter the force or skill however.

Ready for review

emmanuelbassil commented:
Added new puncture var; as having utensils/pens be sharp was deemed unbalanced.
Sharp items, items with the puncture var, and lit welders/cigarettes can now puncture. The latter was unchanged as can_puncture also handles deflating inflatables. 

Ready for review again.

emmanuelbassil commented:
Make it damage glasses/mask instead of destroying them. Hats covering the EYE body part remain indestructible; and protect underlying glasses/mask.
No icon changes; only item name and description.

SierraKomodo commented on code/_helpers/unsorted.dm:
```diff
@@ -901,31 +901,20 @@ GLOBAL_LIST_INIT(duplicate_object_disallowed_vars, list(
 
-//For items that can puncture e.g. thick plastic but aren't necessarily sharp
-//Returns 1 if the given item is capable of popping things like balloons, inflatable barriers, or cutting police tape.
+///For items that can puncture e.g. thick plastic but aren't necessarily sharp.
+///Returns 1 if the given item is capable of popping things like balloons, inflatable barriers, or cutting police tape. Also used to determine what items can eyestab.
 /obj/item/proc/can_puncture()
```
Use the multi-line format for multi-line documentation. Also, the return value should probably be noted as `TRUE` instead of `1`.


```suggestion
/**
 * For items that can puncture e.g. thick plastic but aren't necessarily sharp.
 * 
 * Returns TRUE if the given item is capable of popping things like balloons, inflatable barriers, or cutting police tape. Also used to determine what items can eyestab.
 */
/obj/item/proc/can_puncture()
```

SierraKomodo commented on code/_helpers/unsorted.dm:
```diff
@@ -901,31 +901,20 @@ GLOBAL_LIST_INIT(duplicate_object_disallowed_vars, list(
-
-/obj/item/shovel/can_puncture() //includes spades
-	return 1
+	return welding //Only if welding tool is on.
```
```suggestion
	return welding
```

SierraKomodo commented on code/game/objects/items/weapons/material/misc.dm:
```diff
@@ -56,7 +56,7 @@
 	force_multiplier = 0.2 // 12 with hardness 60 (steel)
 	thrown_force_multiplier = 0.75 // 15 with weight 20 (steel)
-	w_class = ITEM_SIZE_SMALL
+	w_class = ITEM_SIZE_NORMAL
```
New value is identical to parent.


```suggestion
```

SierraKomodo commented on code/game/objects/items.dm:
```diff
@@ -41,10 +41,11 @@
 
-	//This flag is used to determine when items in someone's inventory cover others. IE helmets making it so you can't see glasses, etc.
-	//It should be used purely for appearance. For gameplay effects caused by items covering body parts, use body_parts_covered.
+	///This flag is used to determine when items in someone's inventory cover others. IE helmets making it so you can't see glasses, etc.
+	///It should be used purely for appearance. For gameplay effects caused by items covering body parts, use body_parts_covered.
 	var/flags_inv = 0
```
Multi-line documentation

SierraKomodo commented on code/game/objects/items.dm:
```diff
@@ -41,10 +41,11 @@
 	var/flags_inv = 0
-	var/body_parts_covered = 0 //see setup.dm for appropriate bit flags
+	///See items_clothing.dm for appropriate bit flags
+	var/body_parts_covered = 0
```
```suggestion
	var/body_parts_covered = EMPTY_BITFLAG
```

SierraKomodo commented on code/game/objects/items.dm:
```diff
@@ -579,61 +580,66 @@ var/global/list/slot_flags_enumeration = list(
+	user.do_attack_animation(M) //Addition of fingerprints is handled in post_use_item, since this is called by use_weapon. By returning TRUE is goes to post_use item.
+	playsound(loc, hitsound, 50, TRUE, -1)
+
+	for (var/obj/item/protection in list(H.head, H.wear_mask, H.glasses)) //Order in list is important, it will check head and mask first. If protective hat covers masks/eyes; they can't be destroyed.
+		var/breakcover =  force * user.get_skill_value(SKILL_COMBAT) //Make it impossible to break glasses with screwdrivers/pens (force 2-3); no matter how skilled.
+		if (protection != H.head && (protection.body_parts_covered & EYES) && breakcover > 20) //Only make eye/mask items destroyable; else riot helmets can be destroyed with eye stabbing.
+			H.visible_message(SPAN_DANGER("\The [M]'s [protection] has been damaged with by \the [src] by \the [user]."))
+			CLEAR_FLAGS(protection.body_parts_covered, EYES)
+			protection.name = "Damaged [protection.name]"
+			protection.desc += "<br>[SPAN_NOTICE("It looks damaged.")]"
+			H.unEquip(protection, H.loc)
+			continue
+		if (protection && (protection.body_parts_covered & EYES))
```
This has the potential of damaging 3 worn items and the eyes, _and_ displaying multiple lines of messages as a result.

emmanuelbassil replied:
Yeah, changed to just make it return TRUE if it breaks an item.
So the item dies protecting your eyes, if wearing protective mask + glasses will need three successful attacks to land an eye attack. Given the high miss chance of hitting someone's eyes when they're up, attacking eyes is only really viable when someone is down/buckled/aggro grabbed. These sorts of attacks always land. Which is a good thing.

SierraKomodo commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -100,6 +100,17 @@
+		var/hit_zone = resolve_item_attack(weapon, user, user.zone_sel.selecting)
+		if (!hit_zone) //Miss message to user is processed in resolve_item_attack.
+			return TRUE
+		if (hit_zone == BP_HEAD || hit_zone == BP_EYES) //Resolve_item_attack on non-self mobs never returns BP_EYES, but changes it to BP_HEAD because of check_zone()
```
Players going for head attacks instead of eyes may find this unintuitive or frustrating if they're trying to bash someone and are getting blocked by the eyestab calls.

emmanuelbassil replied:
As discussed on discord, this check only runs if a user is attacking eyes.
This is because resolve_item_attack converts eye attacks to head attacks because ranged attacks also use this proc.

SierraKomodo reviewed (changes requested):
I was apparently pushing the wrong button. See above comments for review.

SierraKomodo commented on code/game/objects/items.dm:
```diff
@@ -579,61 +583,66 @@ var/global/list/slot_flags_enumeration = list(
+	for (var/obj/item/protection in list(H.head, H.wear_mask, H.glasses)) //Order in list is important, it will check head first. If protective hat covers masks/eyes; they can't be destroyed.
+		var/breakcover =  force * user.get_skill_value(SKILL_COMBAT) //Make it impossible to break glasses with screwdrivers/pens (force 2-3); no matter how skilled.
+		if (protection != H.head && (protection.body_parts_covered & EYES) && breakcover > 20) //Only make eye/mask items destroyable; else riot helmets can be destroyed with eye stabbing.
+			H.visible_message(SPAN_DANGER("\The [M]'s \the [protection] has been damaged with by \the [src] by \the [user]."))
```
"Urist McAntag's the screwdriver" vs "Urist McAntag's screwdriver"

```suggestion
			H.visible_message(SPAN_DANGER("\The [M]'s [protection.name] has been damaged with by \the [src] by \the [user]."))
```

SierraKomodo commented on code/game/objects/items.dm:
```diff
@@ -579,61 +583,66 @@ var/global/list/slot_flags_enumeration = list(
+		if (protection != H.head && (protection.body_parts_covered & EYES) && breakcover > 20) //Only make eye/mask items destroyable; else riot helmets can be destroyed with eye stabbing.
+			H.visible_message(SPAN_DANGER("\The [M]'s \the [protection] has been damaged with by \the [src] by \the [user]."))
+			CLEAR_FLAGS(protection.body_parts_covered, EYES)
+			protection.name = "Damaged [protection.name]"
```
Atom names should not be capitlized except for proper nouns.

```suggestion
			protection.name = "damaged [protection.name]"
```

SierraKomodo reviewed (changes requested):
Works in testing, just a couple adjustments needed.

emmanuelbassil committed:
Expand eyestab


Fixes


Auto stash before rebase of "refs/heads/vendorcoin"
Fixes


Fixes


Auto stash before rebase of "refs/heads/grabs"


Fixes


Fixes


Fixes

emmanuelbassil commented:
Handled
